### ISSUE TITLE
[ION-1424] feature: Auth component

### ISSUE THREAD

**Issue by fwang** [sst/sst#4893]
Requests:
- https://discord.com/channels/983865673656705025/1177071497974648952/1222581598831120385

<sub>[ION-227](https://linear.app/sst-dev/issue/ION-227/feature-auth-component)</sub>

<sub>[ION-1424](https://linear.app/sst-dev/issue/ION-1424/feature-auth-component)</sub>

**Comment by ryan-hunter-pc** [comment]
@klclee (and anyone else who found this thread trying to get auth working in v3), I was able to get auth working in v3 using a Google OIDC connector.

[Here's a gist where I show how I got it working on my end](https://gist.github.com/ryan-hunter-pc/88e5ecf01cc8cd4c54637361e467be25)

If you're on the SST Discord, I posted [this message which briefly explains on the confusion that many are hitting](https://discord.com/channels/983865673656705025/1289642513258381344/1293327458740998214) (including me, initially).

<details>
<summary>Full explanation:</summary>

The handler created by calling `auth.authorizer(...)` (in v3) provides a singular authorization code flow for your application that sends back a consistent custom JWT session token format that you can define via `createSessionBuilder()`. It can support multiple adapters, but regardless of that, it provides one consistent flow to your application.

The adapter(s) you define in `auth.authorizer(...)` may perform an "inner" OAuth/OIDC flow depending on the adapter, but that is separate from the "outer" authorization code flow provided to your app by the SST authorizer. The confusion I'm seeing around Discord and here seems to be in not understanding that the "inner" flows are conceptually separate from the "outer" flow that you should be using from your application.

v2 auth abstracted all that away nicely by providing the [auth.attach(...)](https://v2.sst.dev/auth#use-the-construct) method, which set up an API endpoint you could link to from your frontend (e.g. `https://your-backend-api.com/auth/google/authorize`), which then handled the authorization code flow from there. Then you could use the [Session.parameter(...)](https://v2.sst.dev/auth#create-a-session) or [Session.cookie(...)](https://v2.sst.dev/auth#create-a-session) methods to pass the resulting custom JWT session token back to your frontend however you wished.

Since v3 does not (yet?) have those abstractions in place, people seem to be trying to link directly to `https://<your-v3-authorizer-url>/google/authorize` from the frontend without any parameters, (incorrectly) assuming that it is a proxy that handles the "outer" authorization code flow like what v2 provided with `auth.attach(...)`.

That's what causes the `Missing redirect_uri` error, because the authorizer handler is expecting the parameters needed to kick off an authorization code flow, and `redirect_uri` is the first of those parameters which gets [validated in the handler](https://github.com/sst/ion/blob/96eccab42ce60a11db3a0c925ab0ba9b9816a54b/sdk/js/src/auth/handler.ts#L277-L280).

[My gist](https://gist.github.com/ryan-hunter-pc/88e5ecf01cc8cd4c54637361e467be25) manually creates a proxy similar to [what v2 would set up for you](https://github.com/sst/sst/blob/71494f261724417c84dab662e6cf43b0645bc2bb/packages/sst/src/constructs/Auth.ts#L187-L253) when you used `auth.attach(...)`. There are some rough edges where I hard-coded things for simplicity, but the result is that (just like in v2), you can link to `https://your-backend-api.com/auth/google` from your frontend, and the proxy will forward that to the authorizer with the proper parameters, perform the authorization code flow, and then callback to your frontend URL with `?token=<token>` where `<token>` is the custom JWT session defined by `createSessionBuilder()`. It can obviously be adapted to a cookie flow or any other similar approach as needed.
</details>

**Comment by ryan-hunter-pc** [comment]
@fwang @thdxr is there anything I can help with on v3 auth?

A couple things that jump to mind that could be helpful:

- Implementing a wrapper/proxy concept similar to what v2 had with [auth.attach(...)](https://v2.sst.dev/auth#use-the-construct) and [Session.parameter(...)](https://v2.sst.dev/auth#create-a-session) / [Session.cookie(...)](https://v2.sst.dev/auth#cookies), since the lack of that in v3 [seems to be a source of common confusion right now](https://discord.com/channels/983865673656705025/1289642513258381344/1293327458740998214)
- Documenting how to get it working as-is, if there is a reason you haven't brought in a wrapper/proxy concept to v3. I already have a head start in [this gist](https://gist.github.com/ryan-hunter-pc/88e5ecf01cc8cd4c54637361e467be25) and [my message above](https://github.com/sst/sst/issues/4893) 

I would normally just go ahead and start a PR, but figured you might already be on it since you have a Linear ticket for it, so i'm checking first...

Let me know if I can help since I already have the context loaded up in my brain RAM, or If you don't want another cook in the kitchen.

I appreciate all that the SST team does, it is an amazing tool!

**Comment by ilkergzlkkr** [comment]
https://openauth.js.org
https://sst.dev/docs/component/aws/auth
https://youtu.be/SSjNUuQ06tk?si=f2G8GqCXC8d8p8mF

**Comment by Meeeeez** [comment]
> https://openauth.js.org https://sst.dev/docs/component/aws/auth https://youtu.be/SSjNUuQ06tk?si=f2G8GqCXC8d8p8mF

This looks promising!

**Comment by jayair** [comment]
Yeah working on the docs and finalizing the API for it.

**Comment by MarkSFrancis** [comment]
Some feedback from an SST newbie in case it's helpful!

**TL;DR** - could we get please some example `authorizer` code in the SST AWS Auth component's docs that explains how it uses OpenAuth's environment variable support to connect by default to DynamoDB, and also some docs for OpenAuth that explains that it uses environment variables by default for unset config options (like storage)? 

If you'd like that, I'm happy to raise some PRs!

---

I personally struggled getting the Auth component set up, particularly as someone new to SST. 

I imported a series of components from `@openauthjs/openauth`, following [the introductory YouTube video](https://www.youtube.com/watch?v=SSjNUuQ06tk) + [OpenAuth docs](https://github.com/openauthjs/openauth/blob/master/README.md), and used SST's [`sst.aws.Auth`](https://sst.dev/docs/component/aws/auth) component to help deploy it all. But I then struggled to work out how to get `openauth`'s handler connected to the Auth component's DynamoDB storage. 

```ts
import { authorizer } from "@openauthjs/openauth";
import { PasswordAdapter } from "@openauthjs/openauth/adapter/password";
import { DynamoStorage } from '@openauthjs/openauth/storage/dynamo';
import { PasswordUI } from "@openauthjs/openauth/ui/password";
import { handle } from "hono/aws-lambda";
import { Resource } from 'sst';

const app = authorizer({
  storage: DynamoStorage({
    // Unclear if I should be setting this at all, or if I'm doing something wrong. Either way, this confused me a lot
    table: Resource.Auth['???'],
  }),
  providers: {
    password: PasswordAdapter(
      PasswordUI({
        sendCode: async (email, code) => {
          console.log(`Send code ${code} to ${email}`);
        },
      })
    ),
  },
  // ...
});

export const handler = handle(app);
```

I then noticed [the example in the SST repo](https://github.com/sst/sst/blob/dev/examples/aws-auth/src/auth.ts), which used a completely different set of imports, so I switched to that:
```ts
import { auth } from "sst/auth";

export const handler = auth.authorizer({
  // ...
});
```
But that had different problems:
1. I still had the `storage` option available, so it's still unclear if it's being set to DynamoDB for me or not
2. It didn't support the `PasswordAdapter` (presumably because the `PasswordAdapter` is newer?)

I was now feeling confused for which version was "correct", and I'm not sure how to proceed. It feels like I'm either importing something that might be out of date + can't do password auth easily, or I need to write the SST infra code again in order to set up the infra. 

Thankfully, I've had a look around both repos + can now see what's happening. OpenAuth can be configured via environment variables, and uses them when the settings aren't configured in the code. These environment variables are what SST's Auth component set, so there's no need for using SST's normal `Link` techniques.

**Comment by jayair** [comment]
@MarkSFrancis Yeah sorry about the lack of docs. It's what we are working on right now for an OpenAuth v1 and Auth component release.

**Comment by MarkSFrancis** [comment]
@jayair I totally understand, it's still in beta after all! Just to be clear - I'm just providing feedback in the hopes of it being useful, it's not at all a complaint. Great work so far 🚀

**Comment by janvdhorst** [comment]
any updates on this? has been three months since the last post. We can't migrate our app to v3 without cognito auth

**Comment by pstovik** [comment]
FYI: In our project which was SST2 based, we have used Auth construct with just a single OAuth provider
- when considering use of OpenAuth for SST3 we preferred not to use it (as it seems to be good for cases with multiple providers not just single one)
- so mostly copied some parts of the "auth construct" code of SST2 + kept underlying library "openid-client" deployed as the standard Http/proxy Lambda (which manages the session JWTs creation)

**Comment by jayair** [comment]
We added support for Auth while back, closing this.

https://sst.dev/docs/component/aws/auth/

**Comment by twiddler** [comment]
@jayair https://sst.dev/docs/component/aws/auth/ says

> Auth and OpenAuth are currently in beta.

Is `Auth` still in beta? If so, what makes it beta?

I'd love to update to v3, but this is holding me back.

Thanks! 🙏
